active-rfcs/0000-finalize-head-body.md

```diff
@@ -0,0 +1,80 @@
+
+## Non-Goals
+
+1. **Out of scope:** Support `<head>` Injection inside of individual components. This is a commonly requested user feature, but can be tackled as an additional feature, leaving this RFC focused on clarifying our existing API. This RFC does not impact that feature request.
+
+
+# Detailed design
+
+## Template Changes
+
+- `<!DOCTYPE html>`
+  - compiler: completely ignored by the compiler, stripped out from render output.
+  - runtime: a `<!DOCTYPE html>` tag is always included in final page HTML output.
+- `<html>`, `<body>`, `<head>`
+  - compiler: will be left as-is in the component template.
+  - runtime: will be left as-is in final page HTML output.
+  - runtime: may warn if duplicate `html`, `body`, and `head` tags are included in final page HTML output.
+
+## Compiler Changes
+
+- `src/pages` and `src/layouts` will no longer be parsed or rendered differently from other components
+- `as?: 'document' | 'fragment';` will be removed from the compiler.
+- All Astro components will now be compiled as fragments.
+
+Losing `"as": "document"` parse mode will remove some special parser handling, like getting an implicit `<head>` to wrap a standalone `<title>` element. This is intentional to bring us more inline with the RFC, where we respect the HTML authored by the developer as the source of truth and leave mostly as-is in the final output.
+
+In this design it is more "on you" to write valid HTML. This comes from the reality that an imported component can contain unknown HTML, so the compiler can't implicitly assume anything about what is or is not included included the final parent component template.  See https://github.com/withastro/astro/issues/2022 for examples of when this breaks down today. We can help with some static linting, and runtime warnings if the final HTML output is invalid. However, this RFC acknowledges the reality that already exists in v0.21 that imported components break any assumptions and help that we previously attempted to provide.
+
+# Drawbacks
+
+- This RFC will allow you to author a page of HTML that does not actually output `<html>`, `<body>` or `<head>` elements. This can be considered an advantage in that it is more flexible than something like Svelte or Vue. However, it means we need to be more diligent with testing this kind of output across our codebase and dependencies. For example, we would need to confirm that Vite does not have trouble injecting things into an HTML document without a head or body.
+  - If we are blocked by this limitation in some way and can't reasonably unblock (for example, a non-trivial Vite limitaiton) this RFC states that it would be acceptable for Astro to add or require a `<head>` or `<body>` element in the final output HTML in a way that keeps the output HTML valid but unblocks Astro/tooling. Reminder that it is not an explicit goal to support authoring with `<head>` or `<body>`, but it is something that we would like to support if we reasonably can.
```

Comment:
cc @jonathantneal @matthewp making sure you get eyes on this point!

Reply:
Not a goal to support authoring *without* I think you mean?

Reply:
Agreed it’s meant to read as '_without_', and agreed on the sentiment.

I’m unclear on the limitations right now, which IMHO put other parts of this RFC in more uncertainty than this bit. :)

Reply:
rewrote this as a part the latest changes, apologies for the typo!